katcp/testutils.py

```diff
@@ -2049,3 +2051,41 @@ def _advance(self, future):
         yield self.test_instance.set_ioloop_time(
             self.test_instance.ioloop_time + self.quantum)
         future.set_result(None)
+
+
+@contextmanager
+def assert_no_memory_leaks():
+    """Context handler that asserts no memory is leaking within.
+
+    Example
+    -------
+
+    Creating object, then releasing reference:
+
+        >>> with assert_no_memory_leaks():
+        >>>     my_obj = MyClass()
+        >>>     my_obj.optionally_do_stuff()
+        >>>     my_obj = None
+
+    Raises
+    ------
+    AssertionError
+        If the garbage collector finds any unreachable objects when
+        exiting the context handler.
+
+    """
+    old_config = gc.get_debug()
+    gc.set_debug(gc.DEBUG_SAVEALL)
+    try:
+        # check for initial garbage (e.g., test framework may have other leaks)
+        gc.collect()
+        initial_garbage = set(id(g) for g in gc.garbage)
+        yield
+        # collect again and see if there is any new garbage
+        gc.collect()
+        final_garbage = set(id(g) for g in gc.garbage)
+        new_garbage = final_garbage.difference(initial_garbage)
+        unreachable = len(new_garbage)
+        assert unreachable == 0, "{} unreachable objects found".format(unreachable)
```

Comment:
Using `assert` rather than the unittest assertions means that the assert will disappear if run with `python -O` (unless. nose.tools provides the unittest assertions outside of the TestCase class e.g. `nose.tools.assert_equal(0, unreachable)`.

Reply:
@bmerry sorry to sidetrack you, but when would you need to run `python -O`?

Reply:
You might want to verify that the system under test functions correctly with `-O`. I vaguely recall that some project I contributed to (possibly dask) does this as part of CI because there had been such a bug at one point.

It could also happen accidentally if someone sets `PYTHONOPTIMIZE` in their environment in an attempt to speed up all their Python programs.

Reply:
Interesting that you mention that - one of my earlier iterations did pass in the test case object, so I can go back to that.  I thought using a vanilla `assert` would be more generally useful, but the only place we really want to use this context manager is in our own tests.